### PR TITLE
Added sigint protection while entering/leaving cmd2 and Python interactive shells

### LINKED ISSUE #686: Add more sigint protection
Add some more sigint protection around blocks of code that alter or restore critical state data of cmd2. This includes blocks that edit or restore tab completion settings like in `_cmdloop()` and `do_py()`.

### PR DESCRIPTION
**Pull request by kmvanbrunt**
Fixes #686
This should prevent getting into mixed environment states by allowing environment changes to occur atomically.

### CODE REVIEW

File: cmd2/cmd2.py [hunk 1]
```diff
@@ -280,6 +281,24 @@ def cmd_wrapper(cmd2_instance, statement: Union[Statement, str]):
     return arg_decorator
 
 
+class _SavedReadlineSettings:
```
Comment (tleonhardt) [inline cmd2/cmd2.py]: Does this code belong in rl_utils.py?

File: cmd2/cmd2.py [hunk 2]
```diff
@@ -280,6 +281,24 @@ def cmd_wrapper(cmd2_instance, statement: Union[Statement, str]):
...
+    """readline settings that are backed up when switching between readline environments"""
+    def __init__(self):
+        self.completer = None
+        self.delims = ''
+        self.basic_quotes = None
+
+
+class _SavedCmd2Env:
```
Comment (tleonhardt) [inline cmd2/cmd2.py]: Should this be in either utils.py or in pyscript_bridge.py?

File: cmd2/cmd2.py [hunk 3]
```diff
@@ -38,6 +38,7 @@
 import re
 import sys
 import threading
+from code import InteractiveConsole
```
Comment (tleonhardt) [inline cmd2/cmd2.py]: Lots of new code is added here, but no new unit tests.  Should a few unit tests be added?

File: cmd2/cmd2.py (hunk 1 above)
Comment (kmvanbrunt) [reply cmd2/cmd2.py]: Since this class is only used in cmd2.py, I'm OK leaving it.

File: cmd2/cmd2.py (hunk 2 above)
Comment (kmvanbrunt) [reply cmd2/cmd2.py]: Since this class is only used in cmd2.py, I'm OK leaving it.

File: cmd2/cmd2.py (hunk 3 above)
Comment (kmvanbrunt) [reply cmd2/cmd2.py]: I really just moved existing code into functions. No unit test or coverage changes.

**Comment by kmvanbrunt** [comment]
@tleonhardt No need for CHANGELOG updates on this

**Pull request merged**